cmd/zsysd/client/state.go

```diff
@@ -286,3 +290,28 @@ func removeStateGRPC(client *zsys.ZsysLogClient, force, dryrun, system bool, use
 
 	return err
 }
+
+func validateStateName(stateName string) error {
+	var invalidChars []string
+
+	// List of valid characters from zcommon->zfs_namecheck->valid_char()
+	// Space is also valid but not supported by grub and init so booting from a snapshot with a space in the name fails
+	if string(stateName[0]) == "-" ||
+		string(stateName[0]) == "." {
+		return errors.New(i18n.G("state name cannot start with '.', '..' or '-'"))
+	}
+
+	for _, c := range stateName {
+		if !((c >= 'a' && c <= 'z') ||
+			(c >= 'A' && c <= 'Z') ||
+			(c >= '0' && c <= '9') ||
+			c == '-' || c == '_' || c == '.' || c == ':') {
+			invalidChars = append(invalidChars, "'"+string(c)+"'")
```

Comment:
I find it hard to read, maybe, as it’s an error case and performance won’t be important in that one, use fmt.Sprintf?